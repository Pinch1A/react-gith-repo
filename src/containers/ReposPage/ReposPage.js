import * as actions from '../../store/actions/index'
import React, { useEffect, useState } from 'react'
import { RepoList, SearchForm, Spinner } from '../../components'
import { GET_ORGANIZATION } from '../../api/gitV4'
import { connect } from 'react-redux'
import styled from 'styled-components'

const RepoPage = styled.div`
  padding: 20px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
`

const ReposPage = ({
  repos,
  token,
  loading,
  repoSearch,
  pageInfo,
  error,
  clearSearch,
}) => {
  const [repoName, setRepoName] = useState('')

  useEffect(() => {
    const queryString = `is:public ${repoName} in:name`
    clearSearch()
    repoSearch(GET_ORGANIZATION(queryString), token)
  }, [clearSearch, repoName, repoSearch, token])

  const handleForm = (e) => {
    e.preventDefault()
    const queryString = `is:public ${repoName} in:name`
    clearSearch()
    repoSearch(GET_ORGANIZATION(queryString), token)
  }
  const inputChangeHandler = (target) => {
    // event.preventDefault()
    setRepoName(target)
  }

  const fetchMoreRepo = () => {
    let { endCursor } = pageInfo
    let queryString = `is:public ${repoName} in:name`
    return repoSearch(GET_ORGANIZATION(queryString, endCursor), token)
  }

  return (
    <RepoPage id="top">
      {error ? (
        <p> {error.message} </p>
      ) : (
        <div>
          <SearchForm
            handleForm={handleForm}
            searchString={repoName}
            onChange={inputChangeHandler}
            clearSearch={clearSearch}
          />
          <p className="searchCount">
            Search Result: Total {repos ? repos.length : 0} Repository
          </p>
          {repos && !loading && (
            <RepoList
              repos={repos}
              fetchMoreRepo={fetchMoreRepo}
              pageInfo={pageInfo}
            />
          )}
          {loading && <Spinner />}
        </div>
      )}
    </RepoPage>
  )
}

const mapStateToProps = (state) => {
  return {
    repos: state.repoReducer.repos,
    loading: state.repoReducer.loading,
    error: state.repoReducer.error,
    token: state.tokenValidator.token,
    pageInfo: state.repoReducer.pageInfo,
  }
}
const mapDispatchToProps = (dispatch) => {
  return {
    repoSearch: (query, token) => actions.repoSearch(query, token)(dispatch),
    clearSearch: () => dispatch(actions.clearSearch()),
  }
}
export default connect(mapStateToProps, mapDispatchToProps)(ReposPage)
