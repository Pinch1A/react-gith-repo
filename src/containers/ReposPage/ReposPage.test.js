import { mount, shallow } from 'enzyme'
import { Provider } from 'react-redux'
import React from 'react'
import ReposPage from './ReposPage'
import configureStore from 'redux-mock-store'

const mockStore = configureStore([])

const reposMock = [
  {
    node: {
      id: 'id-1',
      name: 'test-2',
      description: 'mock test 1',
      stargazers: { totalCount: 101 },
      url: 'test.test',
    },
  },
  {
    node: {
      id: 'id-2',
      name: 'test-2',
      description: 'mock test 2',
      stargazers: { totalCount: 102 },
      url: 'test.test',
    },
  },
]
const mockReducer = (data) => {
  return {
    repos: reposMock,
    loading: false,
    pageInfo: {},
    error: null,
    ...data,
  }
}

const mockTokenValidator = (data) => {
  return {
    token: 'token-test',
    ...data,
  }
}

describe('<ReposPage>', () => {
  let store

  beforeEach(() => {
    store = mockStore({
      repoReducer: mockReducer(),
      tokenValidator: mockTokenValidator(),
    })
  })
  it('renders without crashing', () => {
    const props = {
      ...mockReducer(),
      ...mockTokenValidator(),
      repoSearch: jest.fn(),
      clearSearch: jest.fn(),
    }

    const wrapper = shallow(
      <Provider store={store}>
        <ReposPage {...props} />
      </Provider>
    )
    expect(wrapper).toMatchSnapshot()
  })
  it('renders error message', () => {
    const error = {
      error: { message: 'Test' },
    }
    const props = {
      ...mockReducer(error),
      ...mockTokenValidator(),
      repoSearch: jest.fn(),
      clearSearch: jest.fn(),
    }

    store = mockStore({
      repoReducer: mockReducer(error),
      tokenValidator: mockTokenValidator(),
    })

    const wrapper = mount(
      <Provider store={store}>
        <ReposPage {...props} />
      </Provider>
    )
    expect(wrapper.find('p').text()).toContain('Test')
  })
  it('renders repos list', () => {
    const props = {
      ...mockReducer(),
      ...mockTokenValidator(),
      repoSearch: jest.fn(),
      clearSearch: jest.fn(),
    }

    store = mockStore({
      repoReducer: mockReducer(),
      tokenValidator: mockTokenValidator(),
    })

    const wrapper = mount(
      <Provider store={store}>
        <ReposPage {...props} />
      </Provider>
    )
    expect(wrapper.find('p').text()).toContain(
      `Search Result: Total ${reposMock.length} Repository`
    )
  })
})
