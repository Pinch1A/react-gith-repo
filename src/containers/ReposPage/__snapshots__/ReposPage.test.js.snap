// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReposPage> renders without crashing 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Connect(ReposPage)
    clearSearch={[MockFunction]}
    error={null}
    loading={false}
    pageInfo={Object {}}
    repoSearch={[MockFunction]}
    repos={
      Array [
        Object {
          "node": Object {
            "description": "mock test 1",
            "id": "id-1",
            "name": "test-2",
            "stargazers": Object {
              "totalCount": 101,
            },
            "url": "test.test",
          },
        },
        Object {
          "node": Object {
            "description": "mock test 2",
            "id": "id-2",
            "name": "test-2",
            "stargazers": Object {
              "totalCount": 102,
            },
            "url": "test.test",
          },
        },
      ]
    }
    token="token-test"
  />
</ContextProvider>
`;
